File: cpp11/Ice/invoke/PrinterI.cpp
//
// Copyright (c) ZeroC, Inc. All rights reserved.
//

#include <Ice/Ice.h>
#include <PrinterI.h>

using namespace std;
using namespace Demo;

static ostream&
operator<<(ostream& out, Color c)
{
    switch(c)
    {
    case Color::red:
        out << "red";
        break;
    case Color::green:
        out << "green";
        break;
    case Color::blue:
        out << "blue";
        break;
    }
    return out;
}

bool
PrinterI::ice_invoke(vector<Ice::Byte> inParams, vector<Ice::Byte>& outParams, const Ice::Current& current)
{
    auto communicator = current.adapter->getCommunicator();

    bool result = true; // success as opposed to user exception

    Ice::InputStream in(communicator, inParams);
    in.startEncapsulation();

    if(current.operation == "printString")
    {
        string message;
        in.read(message);
        cout << "Printing string `" << message << "'" << endl;
    }
    else if(current.operation == "printStringSequence")
    {
        Demo::StringSeq seq;
        in.read(seq);
        cout << "Printing string sequence {";
        for(Demo::StringSeq::iterator p = seq.begin(); p != seq.end(); ++p)
        {
            if(p != seq.begin())
            {
                cout << ", ";
            }
            cout << "'" << *p << "'";
        }
        cout << "}" << endl;
    }
    else if(current.operation == "printDictionary")
    {
        Demo::StringDict dict;
        in.read(dict);
        cout << "Printing dictionary {";
        for(Demo::StringDict::iterator p = dict.begin(); p != dict.end(); ++p)
        {
            if(p != dict.begin())
            {
                cout << ", ";
            }
            cout << p->first << "=" << p->second;
        }
        cout << "}" << endl;
    }
    else if(current.operation == "printEnum")
    {
        Demo::Color c;
        in.read(c);
        cout << "Printing enum " << c << endl;
    }
    else if(current.operation == "printStruct")
    {
        Demo::Structure s;
        in.read(s);
        cout << "Printing struct: name=" << s.name << ", value=" << s.value << endl;
    }
    else if(current.operation == "printStructSequence")
    {
        Demo::StructureSeq seq;
        in.read(seq);
        cout << "Printing struct sequence: {";
        for(Demo::StructureSeq::iterator p = seq.begin(); p != seq.end(); ++p)
        {
            if(p != seq.begin())
            {
                cout << ", ";
            }
            cout << p->name << "=" << p->value;
        }
        cout << "}" << endl;
    }
    else if(current.operation == "printClass")
    {
        shared_ptr<Demo::C> c;
        in.read(c);
        cout << "Printing class: s.name=" << c->s.name << ", s.value=" << c->s.value << endl;
    }
    else if(current.operation == "getValues")
    {
        auto c = make_shared<Demo::C>();
        c->s.name = "green";
        c->s.value = Color::green;
        Ice::OutputStream out(communicator);
        out.startEncapsulation();
        out.write(c);
        out.write("hello");
        out.endEncapsulation();
        out.finished(outParams);
    }
    else if(current.operation == "throwPrintFailure")
    {
        cout << "Throwing PrintFailure" << endl;
        Demo::PrintFailure ex;
        ex.reason = "paper tray empty";
        Ice::OutputStream out(communicator);
        out.startEncapsulation();
        out.write(ex);
        out.endEncapsulation();
        out.finished(outParams);
        result = false;
    }
    else if(current.operation == "shutdown")
    {
        current.adapter->getCommunicator()->shutdown();
    }
    else
    {
        Ice::OperationNotExistException ex(__FILE__, __LINE__);
        ex.id = current.id;
        ex.facet = current.facet;
        ex.operation = current.operation;
        throw ex;
    }

    //
    // Call endEncapsulation to make sure we read all in parameters
    //
    in.endEncapsulation();
    return result;
}

File: swift/Ice/invoke/src/Server/PrinterI.swift
//
// Copyright (c) ZeroC, Inc. All rights reserved.
//

import Ice
import Foundation

extension Structure: CustomStringConvertible {
    public var description: String {
        return "name=\(name), value=\(value)"
    }
}

extension C: CustomStringConvertible {
    public var description: String {
        return "s.name=\(s.name), s.value=\(s.value)"
    }
}

class PrinterI: Ice.Blobject {

     func ice_invoke(inEncaps: Data, current: Current) throws -> (ok: Bool, outParams: Data) {

        let communicator = current.adapter!.getCommunicator()

        var inStream: Ice.InputStream!
        if inEncaps.count > 0 {
            inStream = Ice.InputStream(communicator: communicator, bytes: inEncaps)
            try inStream.startEncapsulation()
        }

        switch current.operation {
        case "printString":
            let message: String = try inStream.read()
            try inStream.endEncapsulation()
            print("Printing string `\(message)'")
            return (ok: true, outParams: Data())
        case "printStringSequence":
            let seq: StringSeq = try inStream.read()
            try inStream.endEncapsulation()
            print("Printing string sequence \(seq)")
            return (ok: true, outParams: Data())
        case "printDictionary":
            let dict = try StringDictHelper.read(from: inStream)
            try inStream.endEncapsulation()
            print("Printing dictionary \(dict)")
            return (ok: true, outParams: Data())
        case "printEnum":
            let c: Color = try inStream.read()
            try inStream.endEncapsulation()
            print("Printing enum \(c)")
            return (ok: true, outParams: Data())
        case "printStruct":
            let s: Structure = try inStream.read()
            try inStream.endEncapsulation()
            print("Printing struct: \(s)")
            return (ok: true, outParams: Data())
        case "printStructSequence":
            let seq = try StructureSeqHelper.read(from: inStream)
            try inStream.endEncapsulation()
            print("Printing struct sequence: \(seq)")
            return (ok: true, outParams: Data())
        case "printClass":
            var c: Value?
            try inStream.read { c = $0 }
            try inStream.readPendingValues()
            try inStream.endEncapsulation()
            print("Printing class: \(c!)")
            return (ok: true, outParams: Data())
        case "getValues":
            let c = C(s: Structure(name: "green", value: .green))
            let outStream = Ice.OutputStream(communicator: communicator)
            outStream.startEncapsulation()
            outStream.write(c)
            outStream.write("hello")
            outStream.writePendingValues()
            outStream.endEncapsulation()
            return (ok: true, outParams: outStream.finished())
        case "throwPrintFailure":
            print("Throwing PrintFailure")
            let ex = PrintFailure(reason: "paper tray empty")
            let outStream = Ice.OutputStream(communicator: communicator)
            outStream.startEncapsulation()
            outStream.write(ex)
            outStream.endEncapsulation()
            return (ok: false, outParams: outStream.finished())
        case "shutdown":
            current.adapter?.getCommunicator().shutdown()
            return (ok: true, outParams: Data())
        default:
            throw Ice.OperationNotExistException(id: current.id,
                                                 facet: current.facet,
                                                 operation: current.operation)
        }
    }
}
